src/metabase/driver/sql_jdbc/actions.clj

```diff
@@ -303,3 +303,57 @@
             [(conj errors {:index row-index, :error (ex-message e)})
              created-rows]))))
      rows)))
+
+(defn- pk-values->filters [table-id pk-values]
+  ;; check that all pk-values have the same shape
+  (when-not (apply = (map (comp set keys) pk-values))
+    (throw (ex-info (trs "Inconsistent row selection.")
+                    {:status-code 400, :pk-values pk-values})))
+  (let [given-pks (set (keys (first pk-values)))
+        table-pks (into {} (map (juxt :name :id) (filter #(= (:semantic_type %) :type/PK) (:fields (first (table/with-fields [(Table table-id)]))))))]
+    (when-not (= given-pks (set (keys table-pks)))
+      (throw (ex-info (trs "Must select with all PKs.")
+                      {:status-code 400, :given-pks (sort given-pks) :table-pks (sort (keys table-pks))})))
+    (map
+      (fn [pk-value]
+        (into [:and]
+              (mapv (fn [[pk-name value]]
+                      [:= [:field (get table-pks pk-name) nil] value])
+                    pk-value)))
+      pk-values)))
+
+(defmethod actions/perform-action!* [:sql-jdbc :bulk/delete]
+  [driver _action {database-id :id, :as database} {:keys [table-id pk-values]}]
+  (log/tracef "Deleting %d rows" (count pk-values))
+  (with-jdbc-transaction [conn database-id]
+    (transduce
+      (m/indexed)
+      (fn
+        ([]
+         [[] []])
+        ([[errors]]
+         ;; if there were any errors throw an Exception with all the error messages.
+         (when (seq errors)
+           (.rollback conn)
+           (throw (ex-info (tru "Error(s) deleting rows.")
+                           {:status-code 400, :errors errors})))
+         ;; if there we no errors then return the created rows.
+         {:success true})
+        ([[errors] [row-index delete-filter]]
+         (try
+           (do-nested-transaction
+             driver
+             conn
+             (fn []
+               (actions/perform-action!*
+                 driver
+                 :row/delete
+                 database
+                 {:database   database-id
+                  :type       :query
+                  :query      {:source-table table-id
+                               :filter delete-filter}})))
+           [errors]
+           (catch Throwable e
+             [(conj errors {:index row-index, :error (ex-message e)})]))))
```

Comment:
This code is really similar to the `bulk/update` code. I think we should refactor the looping/`transduce` logic out into a separate function so we don't have to repeat this boilerplate every time we want to write a bulk action composed of repeated calls to a single-row action inside of a transaction

Reply:
Actually don't worry about this for now. I have something in the `:bulk/update` branch that abstracts this out. Looks like this basically

```clj
(defn- perform-bulk-action-with-repeated-single-row-actions!
  [{:keys [driver database action action-arg-maps error-message error-xform-fn]}]
  (with-jdbc-transaction [conn (u/the-id database)]
    (transduce
     (map-indexed (fn [i row] [i row]))
     (fn
       ([]
        [[] []])
       ([[errors created-rows]]
        ;; if there were any errors throw an Exception with all the error messages.
        (when (seq errors)
          (.rollback conn)
          (throw (ex-info error-message
                          {:status-code 400, :errors errors})))
        ;; if there we no errors then return the created rows.
        {:created-rows created-rows})
       ([[errors created-rows] [row-index arg-map]]
        (try
          (let [result (do-nested-transaction
                        driver
                        conn
                        (fn []
                          (actions/perform-action!* driver action database arg-map)))]
            [errors (conj created-rows (error-xform-fn result))])
          (catch Throwable e
            [(conj errors {:index row-index, :error (ex-message e)})
             created-rows]))))
     action-arg-maps)))

(defmethod actions/perform-action!* [:sql-jdbc :bulk/create]
  [driver _action database {:keys [table-id rows]}]
  (log/tracef "Inserting %d rows" (count rows))
  (perform-bulk-action-with-repeated-single-row-actions!
   {:driver          driver
    :database        database
    :action          :row/create
    :action-arg-maps (for [row rows]
                       {:database   (u/the-id database)
                        :type       :query
                        :query      {:source-table table-id}
                        :create-row row})
    :error-message   (tru "Error(s) inserting rows.")
    :error-xform-fn  :created-row}))
```

(still needs dox tho)

Since this PR will probably land first I'll update `:bulk/delete` to use the new stuff as well

Reply:
Probably going to tweak it a bit more so you can pass a transducer to handle the results transform and error transform

Reply:
I implemented this in https://github.com/metabase/metabase/pull/24036 and have it working e2e for `bulk/create` and the happy-path and error tests are passing with it. I'm using it for `:bulk/update` as well and the happy path tests are passing (still working on the error tests).

If you want to copy that code into your branch and write your bulk action with `perform-bulk-action-with-repeated-single-row-actions!` you can do it. Otherwise I'm happy to update `bulk/delete` once it lands to use it since it abstracts out all the boilerplate